src/renderer/base/FontResource.cpp

```diff
@@ -0,0 +1,285 @@
+        BYTE dfCharSet;
+        WORD dfPixWidth;
+        WORD dfPixHeight;
+        BYTE dfPitchAndFamily;
+        WORD dfAvgWidth;
+        WORD dfMaxWidth;
+        BYTE dfFirstChar;
+        BYTE dfLastChar;
+        BYTE dfDefaultChar;
+        BYTE dfBreakChar;
+        WORD dfWidthBytes;
+        DWORD dfDevice;
+        DWORD dfFace;
+        DWORD dfBitsPointer;
+        DWORD dfBitsOffset;
+        BYTE dfReserved;
+        DWORD dfFlags;
+        WORD dfAspace;
+        WORD dfBspace;
+        WORD dfCspace;
+        DWORD dfColorPointer;
+        DWORD dfReserved1[4];
+        GLYPHENTRY dfCharTable[CHAR_COUNT];
+        CHAR szFaceName[LF_FACESIZE];
+        BYTE bitmaps[1];
+    };
+#pragma pack(pop)
+}
+
+FontResource::FontResource(const gsl::span<const uint16_t> bitPattern,
+                           const til::size sourceSize,
+                           const til::size targetSize,
+                           const size_t centeringHint) :
+    _bitPattern{ bitPattern.begin(), bitPattern.end() },
+    _sourceSize{ sourceSize },
+    _targetSize{ targetSize },
+    _centeringHint{ centeringHint }
+{
+}
+
+void FontResource::SetTargetSize(const til::size targetSize)
+{
+    if (_targetSize != targetSize)
+    {
+        _targetSize = targetSize;
+        _fontHandle = nullptr;
+    }
+}
+
+FontResource::operator HFONT()
+{
+    if (!_fontHandle && !_bitPattern.empty())
+    {
+        _regenerateFont();
+    }
+    return _fontHandle.get();
+}
+
+void FontResource::_regenerateFont()
+{
+    const auto targetWidth = _targetSize.width<WORD>();
+    const auto targetHeight = _targetSize.height<WORD>();
+    const auto charSizeInBytes = (targetWidth + 7) / 8 * targetHeight;
+
+    const DWORD fontBitmapSize = charSizeInBytes * CHAR_COUNT;
+    const DWORD fontResourceSize = sizeof(FONTINFO) + fontBitmapSize;
+
+    auto fontResourceBuffer = std::vector<byte>(fontResourceSize);
+    void* fontResourceBufferPointer = fontResourceBuffer.data();
+    auto& fontResource = *static_cast<FONTINFO*>(fontResourceBufferPointer);
+
+    fontResource.dfVersion = 0x300;
+    fontResource.dfSize = fontResourceSize;
+    fontResource.dfWeight = FW_NORMAL;
+    fontResource.dfCharSet = OEM_CHARSET;
+    fontResource.dfPixWidth = targetWidth;
+    fontResource.dfPixHeight = targetHeight;
+    fontResource.dfPitchAndFamily = FIXED_PITCH | FF_DONTCARE;
+    fontResource.dfAvgWidth = targetWidth;
+    fontResource.dfMaxWidth = targetWidth;
+    fontResource.dfFirstChar = L' ';
+    fontResource.dfLastChar = fontResource.dfFirstChar + CHAR_COUNT - 1;
+    fontResource.dfFace = offsetof(FONTINFO, szFaceName);
+    fontResource.dfBitsOffset = offsetof(FONTINFO, bitmaps);
+    fontResource.dfFlags = DFF_FIXED | DFF_1COLOR;
+
+    // We use the steady_clock to create a unique name for the font.
+    const auto uniqueValue = std::chrono::steady_clock::now().time_since_epoch().count();
+    sprintf_s(fontResource.szFaceName, "WTSOFTFONT%016llX", uniqueValue);

```

Comment:
OK, I've done something like that now. I just went with `faceNameCounter++` rather than the `fetch_add` call, since that seemed more readable, and I'm assuming the relaxed memory order isn't really essential (and it doesn't seem to make any difference to the generated code, at least on x64).